db/memtable.cc

```diff
@@ -412,13 +412,28 @@ InternalIterator* MemTable::NewIterator(const ReadOptions& read_options,
   return new (mem) MemTableIterator(*this, read_options, arena);
 }
 
-InternalIterator* MemTable::NewRangeTombstoneIterator(
+FragmentedRangeTombstoneIterator* MemTable::NewRangeTombstoneIterator(
     const ReadOptions& read_options) {
   if (read_options.ignore_range_deletions || is_range_del_table_empty_) {
     return nullptr;
   }
-  return new MemTableIterator(*this, read_options, nullptr /* arena */,
-                              true /* use_range_del_table */);
+  auto* unfragmented_iter = new MemTableIterator(
+      *this, read_options, nullptr /* arena */, true /* use_range_del_table */);
+  if (unfragmented_iter == nullptr) {
+    return nullptr;
+  }
+  auto fragmented_tombstone_list =
+      std::make_shared<FragmentedRangeTombstoneList>(
+          std::unique_ptr<InternalIterator>(unfragmented_iter),
+          comparator_.comparator);
+
+  SequenceNumber snapshot = kMaxSequenceNumber;
```

Comment:
I think there should be a seqnum associated with the read to prevent concurrently written range tombstones from being visible, which I'm not sure is safe. Check out how we used `GetInternalKeySeqno(key.internal_key())` in the old code.

Reply:
When the data iterator is created with `MemTable::NewIterator`, it just uses the read options provided, and the range tombstone iterator is created the same way. The seqnum used here is only for snapshots -- I assumed in the normal case that the unfragmented iterator created from the memtable wouldn't race with concurrent writes.

Reply:
After our discussion, I learned that the data iterator skips newer entries at the `DBIter` level. Since that can't happen here, the sequence number from the lookup key must be used (it was being used before so the current code is fine).